drv/psc-seq-server/src/main.rs

```diff
@@ -3,48 +3,603 @@
+    let psu_states: [PsuState; PSU_COUNT] = core::array::from_fn(|i| {
+        PsuState::Present(if initial_psu_enabled[i] {
+            ringbuf_entry!((start_time, Trace::FoundEnabled { psu: i as u8 }));
+            PresentState::On
+        } else {
+            // PSU was forced off by our previous incarnation. Schedule it to
+            // turn back on in the future if things clear up.
+            ringbuf_entry!((
+                start_time,
+                Trace::FoundAlreadyDisabled { psu: i as u8 }
+            ));
+            PresentState::Faulted {
+                turn_on_deadline: start_time.saturating_add(FAULT_OFF_MS),
+            }
+        })
+    });
+    let mut psus = psu_states.map(|state| Psu { state });
+
+    // Turn the chassis LED on to indicate that we're alive.
     sys.gpio_set(STATUS_LED);
+    // TODO: if we wanted to kick jefe into a greater-than-A2 state, this'd be
+    // where it happens.
 
-    // We have nothing else to do, so sleep forever via waiting for a message
-    // from the kernel that won't arrive.
+    // Poll things.
+    sys_set_timer(Some(start_time), notifications::TIMER_MASK);
+    let sleep_notifications = all_pin_notifications | notifications::TIMER_MASK;
     loop {
-        sys_recv_notification(0);
+        sys.gpio_irq_control(all_pin_notifications, IrqControl::Enable)
+            .unwrap_lite();
+
+        let present_l_bits = sys.gpio_read(ALL_PSU_PRESENT_L_PINS);
+        let ok_bits = sys.gpio_read(ALL_PSU_PWR_OK_PINS);
+
+        let now = sys_get_timer().now;
+        for i in 0..PSU_COUNT {
+            // Presence signals are active LOW.
+            let present = if present_l_bits & (1 << PSU_PRESENT_L_PINS[i]) == 0
+            {
+                Present::Yes
+            } else {
+                Present::No
+            };
+            // PWR_OK signals are active HIGH.
+            let ok = if ok_bits & (1 << PSU_PWR_OK_PINS[i]) != 0 {
+                Status::Good
+            } else {
+                Status::NotGood
+            };
+            match psus[i].step(now, present, ok) {
+                None => (),
+
+                Some(ActionRequired::EnableMe) => {
+                    ringbuf_entry!((now, Trace::Enabling { psu: i as u8 }));
+                    // Enable the PSU by allowing `ENABLE_L` to float low, by no
+                    // longer asserting high.
+                    sys.gpio_configure_input(
+                        PSU_ENABLE_L_PORT.pin(PSU_ENABLE_L_PINS[i]),
+                        Pull::None,
+                    );
+                }
+                Some(ActionRequired::DisableMe { attempt_snapshot }) => {
+                    if attempt_snapshot {
+                        // TODO snapshot goes here
+                    }
+                    ringbuf_entry!((
+                        now,
+                        Trace::Disabling {
+                            psu: i as u8,
+                            present: attempt_snapshot,
```

Comment:
this is _very_ nitpicky, but it feels like the naming of this variable seems to be a bit in conflict over whether it represents a _fact_ about the state of the world ("is the PSU `present`?') or a _policy_ about whether or not to perform a particular action ("should we `attempt_snapshot`?"). 

it feels like there is a potential for future issues if we were to later change `step` to decide whether or not to attempt a snapshot based on more factors than _just_ whether the PSU is present, without updating this code. if we were to ever return `attempt_snapshot: false` for a present PSU for some reason, the ringbuffer would incorrectly indicate that the PSU is not present.

just a thought.

Reply:
I think it'll be clearer once I finish and merge the PMBus snapshot code, but I may rename it based on your suggestion at that point. The two names are a result of two different use cases:
- It was going to be named `attempt_snapshot` internally to reflect the implementation intent, but then
- I recognized that that name wasn't very useful in a ringbuf and came up with a different one for logging there. (This one's the user-visible one.)